rfc/0001-policy-server-custom-resource.md

```diff
@@ -0,0 +1,426 @@
+    associated with them.
+
+Some of the resources mentioned above are Namespace-based (like Deployment, ConfigMap, Service);
+in the beginning all these resources will be created in the same Namespace where the controller
+operates. We might want to make that tunable in the future, but right now it would just be
+an overkill.
+
+When a `ClusterAdmissionPolicy` resource is created/updated:
+
+  * Find the `PolicyServer` that owns the policy
+  * Trigger the reconciliation loop of the Policy Server. This will lead to an update
+    of the ConfigMap that holds the `policies.yaml` file, and a rollout of the Deployment
+
+
+Special handling must be done for deletion events. Both `PolicyServer` and `ClusterAdmissionPolicy`
+resources must have a finalizer set.
+
+Deletion of a `ClusterAdmissionPolicy`:
+
+  * Thanks to the finalizer, the `ClusterAdmissionPolicy` resource is not yet deleted
+    from etcd. We can hence find the `PolicyServer` that owns the about-to-be-deleted
+    policy. We will use this information later on.
+  * Delete the `ValidatingWebhookConfiguration` or the `MutatingWebhookConfiguration`
+    resource associated with the policy.
+  * Delete the `ClusterAdmissionPolicy` from etcd
+  * Perform the reconciliation loop of the `PolicyServer` that owned the policy
+
+Deletion of a `PolicyServer`:
+
+  * Thanks to the finalizer, the `PolicyServer` resource is not yet deleted
+    from etcd.
+  * Ensure these operations take place. Note well: the order of execution
+    matters
+    * For each policy owned by the `PolicyServer` instance:
+      - Delete the `ValidatingWebhookConfiguration` or the `MutatingWebhookConfiguration`
+        resource associated with the policy
+      - Delete the `ClusterAdmissionPolicy` from etcd
+    * Delete the `PolicyServer` object from etcd
+
+We have to research whether this can be done by the Kubernetes Garbage Collector
+by setting the [owner references](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/#owner-references)
+or whether this has to be handled by us to ensure the right ordering takes place.
+
+
+## Validation and mutation of our own Custom Resources
+
+We must provide a mutation webhook that goes through our `PolicyServer`
+and `ClusterAdmissionPolicy` objects.
+
+This will perform the following operations:
+
+  * `ClusterAdmissionPolicy`:
+    - Set our finalizer
+    - When the user specifies a value for the `policyServer` attribute, ensure
+      there's actually a `PolicyServer` object with that name.
+  * `PolicyServer`:
+    - Set our finalizer
+    - Reject the delete operations against the `PolicyServer` named `default`
+    - Reject update operations that attempt to change the name of the `PolicyServer`.
+      The name of a `PolicyServer` must be immutable, a lot of relationships
+      between Kubernetes resources and the `PolicyServer` are built on that (for example,
+      the ConfigMap, the Service, the ClusterAdmissionPolicy)
+
```

Comment:
We also want to prohibit `PolicyServer` deletion if any `ClusterAdmissionPolicy` is bound to it.

Reply:
Do we? I think this is useful for quick cleanup operations

Reply:
Allowing to delete a `PoilcyServer` when `ClusterAdmissionPolicy` are bound to it means that we will remove the webhook (at least temporarily until recreated, if recreated at all), and although the `ClusterAdmissionPolicy` is persisted during all that time, requests could have been accepted that the policy would have rejected if been active.

I think kubewarden should not try to be smart here, and prioritize safety.

Reply:
On a second thought: you are right. We can do it safely because we won't remove any webhook. Webhooks will fail to reach the service, and depending on their failure policy they will accept or reject.

Sounds good, no need to block the `PolicyServer` deletion as you have said.

Reply:
I think we can discuss more in details what to do once we decide go go ahead :)